Mathlib/NumberTheory/FLT/Three.lean

```diff
@@ -342,6 +364,292 @@ lemma exists_Solution_of_Solution' : ∃ (S₁ : Solution hζ), S₁.multiplicit
+
+/-- Given `S : Solution`, we let `S.w` be any element such that `S.c = λ ^ S.multiplicity * S.w` -/
+private noncomputable def w :=
+  (multiplicity.pow_multiplicity_dvd S.toSolution'.multiplicity_lambda_c_finite).choose
+
+private lemma w_spec : S.c = λ ^ S.multiplicity * S.w :=
+(multiplicity.pow_multiplicity_dvd S.toSolution'.multiplicity_lambda_c_finite).choose_spec
+
+private lemma lambda_not_dvd_w : ¬ λ ∣ S.w := by
+  intro h
+  replace h := mul_dvd_mul_left (λ ^ S.multiplicity) h
+  rw [← w_spec] at h
+  have hh : _ := multiplicity.is_greatest' S.toSolution'.multiplicity_lambda_c_finite
+    (lt_add_one S.multiplicity)
+  rw [pow_succ', mul_comm] at hh
+  exact hh h
+
+private lemma lambda_not_dvd_x : ¬ λ ∣ S.x := fun h ↦ by
+  replace h := mul_dvd_mul_left (λ ^ (3 * S.multiplicity - 2)) h
+  rw [mul_comm, ← x_spec] at h
+  replace h :=
+    mul_dvd_mul (mul_dvd_mul h S.lambda_dvd_a_add_eta_mul_b) S.lambda_dvd_a_add_eta_sq_mul_b
+  simp only [← a_cube_add_b_cube_eq_mul, S.H, w_spec, Units.isUnit, IsUnit.dvd_mul_left] at h
+  rw [← pow_succ', mul_comm, ← mul_assoc, ← pow_succ'] at h
+  have := S.two_le_multiplicity
+  rw [show 3 * multiplicity S - 2 + 1 + 1 = 3 * multiplicity S by omega, mul_pow, ← pow_mul,
+    mul_comm _ 3, mul_dvd_mul_iff_left _] at h
+  · exact lambda_not_dvd_w _ <| hζ.zeta_sub_one_prime'.dvd_of_dvd_pow h
+  · simp [hζ.zeta_sub_one_prime'.ne_zero]
+
+attribute [local instance] IsCyclotomicExtension.Rat.three_pid
+
+private lemma isCoprime_x_y : IsCoprime S.x S.y := by
+  refine isCoprime_of_prime_dvd (not_and.2 (fun _ hy ↦ lambda_not_dvd_y S (by simp [hy]))) ?_
+  intro p hp p_dvd_x p_dvd_y
+  refine lambda_not_dvd_x S ?_
+  rw [← Associated.dvd_iff_dvd_left <| associated_of_dvd_a_add_b_of_dvd_a_add_eta_mul_b S hp ?_ ?_]
+  · exact p_dvd_x
+  · rw [x_spec]
+    exact dvd_mul_of_dvd_right p_dvd_x (λ ^ (3 * S.multiplicity - 2))
+  · convert dvd_mul_of_dvd_right p_dvd_y (η -1) using 1
+    rw [y_spec, coe_eta]
+
+private lemma isCoprime_x_z : IsCoprime S.x S.z := by
+  refine isCoprime_of_prime_dvd (not_and.2 (fun _ hz ↦ lambda_not_dvd_z S (by simp [hz]))) ?_
+  intro p hp p_dvd_x p_dvd_z
+  refine lambda_not_dvd_x S ?_
+  rw [← Associated.dvd_iff_dvd_left <|
+    associated_of_dvd_a_add_b_of_dvd_a_add_eta_sq_mul_b S hp ?_ ?_]
+  · exact p_dvd_x
+  · rw [x_spec]
+    exact dvd_mul_of_dvd_right p_dvd_x (λ ^ (3 * S.multiplicity - 2))
+  · convert dvd_mul_of_dvd_right p_dvd_z (η - 1) using 1
+    rw [z_spec, coe_eta]
+
+private lemma isCoprime_y_z : IsCoprime S.y S.z := by
+  refine isCoprime_of_prime_dvd (not_and.2 (fun _ hz ↦ lambda_not_dvd_z S (by simp [hz]))) ?_
+  intro p hp p_dvd_y p_dvd_z
+  refine lambda_not_dvd_y S ?_
+  rw [← Associated.dvd_iff_dvd_left <|
+    associated_of_dvd_a_add_eta_mul_b_of_dvd_a_add_eta_sq_mul_b S hp ?_ ?_]
+  · exact p_dvd_y
+  · rw [y_spec]
+    exact dvd_mul_of_dvd_right p_dvd_y (η - 1)
+  · convert dvd_mul_of_dvd_right p_dvd_z (η - 1) using 1
+    rw [z_spec, coe_eta]
+
```

Comment:
The preceding three proofs are all very similar. Can some common argument be extracted, or are they too different for that?

Reply:
They're now very short via `isCoprime_helper`. Thanks for the remark, I wouldn't had the energy to do that otherwise, but it is definitely better.